src/Pact/Types/Term.hs

```diff
@@ -208,26 +215,92 @@ newtype TableName = TableName Text
     deriving (Eq,Ord,IsString,ToTerm,AsString,Hashable)
 instance Show TableName where show (TableName s) = show s
 
-data Module = Module {
-    _mName :: !ModuleName
-  , _mKeySet :: !KeySetName
-  , _mMeta :: !Meta
-  , _mCode :: !Code
-  , _mHash :: !Hash
-  , _mBlessed :: !(HS.HashSet Hash)
-  } deriving (Eq)
+newtype ModuleName = ModuleName Text
+    deriving (Eq,Ord,IsString,ToJSON,FromJSON,AsString,Hashable,Pretty)
+instance Show ModuleName where show (ModuleName s) = show s
+
+-- | A named reference from source.
+data Name =
+    QName { _nQual :: ModuleName, _nName :: Text, _nInfo :: Info } |
+    Name { _nName :: Text, _nInfo :: Info }
+         deriving (Generic)
+instance Show Name where
+    show (QName q n _) = asString' q ++ "." ++ unpack n
+    show (Name n _) = unpack n
+instance ToJSON Name where toJSON = toJSON . show
+instance FromJSON Name where
+  parseJSON = withText "Name" $ \t -> case AP.parseOnly (parseName def) t of
+    Left s -> fail s
+    Right n -> return n
+
+parseName :: (TokenParsing m, Monad m) => Info -> m Name
+parseName i = do
+  a <- ident style
+  try (qualified >>= \qn -> return (QName (ModuleName a) qn i) <?> "qualified name") <|>
+    return (Name a i)
+
+
+instance Hashable Name where
+  hashWithSalt s (Name t _) = s `hashWithSalt` (0::Int) `hashWithSalt` t
+  hashWithSalt s (QName q n _) = s `hashWithSalt` (1::Int) `hashWithSalt` q `hashWithSalt` n
+instance Eq Name where
+  (QName a b _) == (QName c d _) = (a,b) == (c,d)
+  (Name a _) == (Name b _) = a == b
+  _ == _ = False
+instance Ord Name where
+  (QName a b _) `compare` (QName c d _) = (a,b) `compare` (c,d)
+  (Name a _) `compare` (Name b _) = a `compare` b
+  Name {} `compare` QName {} = LT
+  QName {} `compare` Name {} = GT
+  
+
+
+data Module
+ = Module
```

Comment:
See comment elsewhere, maybe a TODO here to move to a safe ADT

Reply:
Fixed